### PR TITLE
Add a post on deploying vision models on Vertex AI

### PR DESCRIPTION
**Pull request by sayakpaul**
@stevhliu, this is the final post in the series on deploying vision models. Here are the previous two posts:

* [Deploying TensorFlow Vision Models in Hugging Face with TF Serving](https://huggingface.co/blog/tf-serving-vision)
* [Deploying 🤗 ViT on Kubernetes with TF Serving](https://huggingface.co/blog/deploy-tfserving-kubernetes) (with [Chansung Park](https://github.com/deep-diver))

@Rocketknight1 @gante would appreciate your reviews on this too. TensorFlow experience should suffice, but GCP experience is nice, not a must-have. 

Done with @deep-diver.

### CODE REVIEW

File: deploy-vertex-ai.md [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+|:----------------------:|:--------------------------:|
+| NVIDIA_TESLA_T4      | $ 0.4024                |
+</div>
+
+# Call for Action
+
+The collection of TensorFlow vision models in 🤗 Transformers is growing. It now has support
+state-of-the-art semantic segmentation with 
+[SegFormer](https://huggingface.co/docs/transformers/model_doc/segformer#transformers.TFSegformerForSemanticSegmentation).
+We encourage you to extend the deployment workflow you learned this post to semantic segmentation models like SegFormer.
```
Comment (sayakpaul) [inline deploy-vertex-ai.md:506]: @deep-diver FYI.

**Review by gante — COMMENTED** [review]
Looks good! I like the call to action 🚀 

Can you share a link for the colab, for easier review?

**Comment by gante** [comment]
Also cc @merveenoyan (TF blog post)

**Comment by gante** [comment]
(Back in my time, one had to do all this manually with gRPC services 🧓 )

**Comment by sayakpaul** [comment]
> Can you share a link for the colab, for easier review?

@gante here you go:
https://colab.research.google.com/github/sayakpaul/deploy-hf-tf-vision-models/blob/main/hf_vision_model_vertex_ai/hf_tf_vision_vertex_ai.ipynb

Thanks for reviewing! 

> (Back in my time, one had to do all this manually with gRPC services 🧓 )

I know some of that pain 😖

**Comment by deep-diver** [comment]
@sayakpaul 

looks like `vertex-ai-serving/notebooks/112_vertex_ai_vision.ipynb` is mistakenly pushed?

**Comment by sayakpaul** [comment]
> @sayakpaul
> 
> looks like `vertex-ai-serving/notebooks/112_vertex_ai_vision.ipynb` is mistakenly pushed?

@deep-diver should be fixed now.

File: deploy-vertex-ai.md [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+to manage Endpoints, Vertex AI lets you easily control the important
+aspects of the deployment.
+
+It takes about 15 - 30 minutes for Vertex AI to scope the deployment.
+Once it’s done, you should be able to see it on the
+[<u>console</u>](https://console.cloud.google.com/vertex-ai/endpoints).
+
+# Performing Predictions
+
+Assuming you succeeded in the above step, you can test the deployed
```
Comment (stevhliu) [inline deploy-vertex-ai.md:328]:
```suggestion
If your deployment was successful, you can test the deployed
```

File: deploy-vertex-ai.md [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+image_path = tf.keras.utils.get_file(
+    "image.jpg", "http://images.cocodataset.org/val2017/000000039769.jpg"
+)
+bytes = tf.io.read_file(image_path)
+b64str = base64.b64encode(bytes.numpy()).decode("utf-8")
+```
+
+**4)** The following utility first prepares a list of instances (only
+one instance in this case) and then uses a prediction service client (of
+type [`PredictionServiceClient`](https://cloud.google.com/python/docs/reference/automl/latest/google.cloud.automl_v1beta1.services.prediction_service.PredictionServiceClient)).
```
Comment (stevhliu) [inline deploy-vertex-ai.md:346]:
```suggestion
one instance in this case) and then uses a [`PredictionServiceClient`](https://cloud.google.com/python/docs/reference/automl/latest/google.cloud.automl_v1beta1.services.prediction_service.PredictionServiceClient)).
```

File: deploy-vertex-ai.md [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+the request payload for each instance needs to be like so:
+
+```py
+{serving_input: {"b64": base64.b64encode(jpeg_data).decode()}}
+```
+
+The `predict_image()` utility prepares the request payload conforming
+to this specification.
+
+If everything goes well with the deployment, if you call
```
Comment (stevhliu) [inline deploy-vertex-ai.md:384]:
```suggestion
If everything goes well with the deployment, when you call
```

File: deploy-vertex-ai.md [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+
+# Local Load Testing
+
+We conducted a local load test to better understand the limits of the
+Endpoint with [<u>Locust</u>](https://locust.io/). The table below
+summarizes the request statistics:
+
+![](./assets/97_vertex_ai/image5.png)
+
+Among all the different statistics shown in the table \`Average (ms)\`
```
Comment (stevhliu) [inline deploy-vertex-ai.md:454]:
```suggestion
Among all the different statistics shown in the table, \`Average (ms)\`
```

File: deploy-vertex-ai.md [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+refers to the average latency of the Endpoint. Locust fired about
+**17230 requests**, and their average latency is reported to be **646
+Milliseconds** which is impressive. In practice, you’d want to simulate
```
Comment (stevhliu) [inline deploy-vertex-ai.md:457]:
```suggestion
refers to the average latency of the Endpoint. Locust fired off about
**17230 requests**, and the reported average latency is **646
Milliseconds**, which is impressive. In practice, you’d want to simulate
```

File: deploy-vertex-ai.md [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+Refer [<u>here</u>](https://cloud.google.com/architecture/load-testing-and-monitoring-aiplatform-models)
+to learn more.
+
+[<u>This directory</u>](https://github.com/sayakpaul/deploy-hf-tf-vision-models/tree/main/hf_vision_model_vertex_ai/locust)
+has all the information needed to know how we conducted the load test.
+
+# Pricing
+
+We used the [<u>GCP cost estimator</u>](https://cloud.google.com/products/calculator) for this
+purpose, and the exact pricing table can be found [<u>here</u>](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).
```
Comment (stevhliu) [inline deploy-vertex-ai.md:468]: I would clarify what `this purpose` refers to since it isn't very clear.

File: deploy-vertex-ai.md [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+<div align="center">
+
+| **Machine Type**            | **Hourly Pricing (USD)** |
+|:-----------------------------:|:--------------------------:|
+| n1-standard-8 (8vCPU, 30GB) | $ 0.4372                |
+
+</div>
+
+Also, when you attach accelerators to the compute node, you will get
+extra charges by the type of the accelerator you want. We have used
```
Comment (stevhliu) [inline deploy-vertex-ai.md:489]:
```suggestion
Also, when you attach accelerators to the compute node, you will be
charged extra by the type of accelerator you want. We used
```

File: deploy-vertex-ai.md [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+
+
+| **Accelerator Type** | **Hourly Pricing (USD)** |
+|:----------------------:|:--------------------------:|
+| NVIDIA_TESLA_T4      | $ 0.4024                |
+</div>
+
+# Call for Action
+
+The collection of TensorFlow vision models in 🤗 Transformers is growing. It now has support
```
Comment (stevhliu) [inline deploy-vertex-ai.md:504]:
```suggestion
The collection of TensorFlow vision models in 🤗 Transformers is growing. It now supports
```

File: deploy-vertex-ai.md [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+state-of-the-art semantic segmentation with 
+[SegFormer](https://huggingface.co/docs/transformers/model_doc/segformer#transformers.TFSegformerForSemanticSegmentation).
+We encourage you to extend the deployment workflow you learned this post to semantic segmentation models like SegFormer.
```
Comment (stevhliu) [inline deploy-vertex-ai.md:507]:
```suggestion
We encourage you to extend the deployment workflow you learned in this post to semantic segmentation models like SegFormer.
```

File: deploy-vertex-ai.md [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+
+# Conclusion
+
+In this post, you learned how to deploy a Vision Transformer model using
```
Comment (stevhliu) [inline deploy-vertex-ai.md:511]:
```suggestion
In this post, you learned how to deploy a Vision Transformer model with
```

File: deploy-vertex-ai.md [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+the Vertex AI platform using the easy APIs it provides. You also learned
+how Vertex AI’s features benefit the model deployment process enabling
+you to focus on declarative configurations while taking away complex
```
Comment (stevhliu) [inline deploy-vertex-ai.md:514]:
```suggestion
how Vertex AI’s features benefit the model deployment process by enabling
you to focus on declarative configurations and removing the complex
```

**Review by stevhliu — COMMENTED** [review]
Very nice call to action at the end! Considering this is the final post in the series, I'd maybe add a brief summary (can just be 2-3 sentences) of what users learned in the first two posts just to tie everything together.

**Comment by sayakpaul** [comment]
@stevhliu we addressed all the comments.

**Comment by stevhliu** [comment]
Looks good to me, I'll merge it tomorrow if there are no other comments :)

**Comment by sayakpaul** [comment]
> If it's ok I would like to give a review tomorrow :)

Please do.

File: deploy-vertex-ai.md [hunk 13]
```diff
@@ -0,0 +1,528 @@
...
+artifacts in a [<u>Google Cloud Storage (GCS) bucket</u>](https://cloud.google.com/storage/docs/json_api/v1/buckets).
+The accompanying Colab Notebook shows how to create a GCS bucket and
+save the model artifacts into it.
+
+# Deployment workflow with Vertex AI 
+
+The figure below gives a pictorial workflow of deploying an already
+trained TensorFlow model on Vertex AI.
+
+![](./assets/97_vertex_ai/image7.png)
```
Comment (osanseviero) [inline deploy-vertex-ai.md:137]: I really like the images in this blog post!

File: deploy-vertex-ai.md [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+
+- Model monitoring window length and its objectives
+
+- Prediction requests sampling rate
+
+# Performing the Deployment
+
+The [`google-cloud-aiplatform`](https://pypi.org/project/google-cloud-aiplatform/)
+Python SDK provides easy APIs to manage the lifecycle of a deployment on
+Vertex AI. It is divided into four steps - **1)** uploading a model,
```
[the remark this answers is not in the document]
Comment (osanseviero) [reply deploy-vertex-ai.md:181]: +1 on making this a list

File: deploy-vertex-ai.md [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+
+# Pricing
+
+You can use the [<u>GCP cost estimator</u>](https://cloud.google.com/products/calculator) to estimate the cost of usage, 
+and the exact hourly pricing table can be found [<u>here</u>](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).
+It is worth noting that you are only charged when the node is processing
+the actual prediction requests, and you need to calculate the price with
+and without GPUs.
+
+For the Vertex Prediction for a custom-trained model, we can choose
```
Comment (osanseviero) [inline deploy-vertex-ai.md:468]: Pricing tends to be one of the things that can change, so I would suggest to keep this section a bit shorter/less precise and instead link more to external resources so this does not become outdated

**Review by osanseviero — APPROVED** [review]
Cool blog post! Thanks a lot for it!

**Comment by deep-diver** [comment]
@osanseviero @sayakpaul 

addressed comments, especially:
- pricing section is now shorter as suggested (removed too precise pricing like hourly price of each vCPU core)
- each step is arranged in list format
- updated publication date

**Comment by osanseviero** [comment]
Awesome, this is great! Feel free to resolve the conflicts and merge afterwards. Please also update the date in _blog.yml

**Comment by deep-diver** [comment]
@osanseviero 

phew.. conflict is resolved, and I updated the date in `_blog.yml` as well

**Comment by osanseviero** [comment]
Awesome blog post, great work! :fire: this is now deployed https://huggingface.co/blog/deploy-vertex-ai

**Pull request merged**